### PR TITLE
Add IoU metric for bounding boxes.

### PR DESCRIPTION
**Pull request by kartikdutt18**
Hey everyone,
This PR aims to add IoU metric to this repo. I need to add more detailed tests for it. Also some metrics such ip_metric don't have tests in `metric_tests.cpp`. Did I miss something?

Also, This metric is for bounding boxes which I would need for NMS and object detection algorithm. I do plan on opening another PR to add mIoU to take as input a matrix where each pixel is label. That is more useful for segmentation so I have left that for now.

Files changed:
- metrics/CMakeLists.txt
- metrics/iou_metric.hpp
- metrics/iou_metric_impl.hpp
- tests/metric_tests.cpp

Changes made :
- Added IoU metric (only for bounding boxes).
- Added a simple test. (will add more soon).

Let me know if I missed something. 
Thanks a ton! 
Regards.

### CODE REVIEW

File: src/mlpack/core/metrics/iou_metric.hpp [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+ *
+ * For bounding box representation there are two common representation
+ * either as coordinates i.e. each value in vector represents a
+ * coordinate in the format x0, y0, x1, y1 where x0, y0 represent the
+ * lower left coordinate and x1, y1, represent upper
+ * right coordinate.
+ * 
+ * Second representation follows the following representation : x0, y0, h, w.
+ * Where bx and by are bottom left bounding box coordinates and h, w are
+ * height and width of the bounding box.
```
Comment (kartikdutt18) [inline src/mlpack/core/metrics/iou_metric.hpp:32]:
```suggestion
  * lower left coordinate and x1, y1, represent upper right coordinate.
  * 
  * Second representation follows the following representation : x0, y0, h, w.
  * Where x0 and x1 are bottom left bounding box coordinates and h, w are
  * height and width of the bounding box.
```
This is a neat trick :)

File: src/mlpack/core/metrics/iou_metric_impl.hpp [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+template<bool useCoordinates>
+template <typename VecTypeA, typename VecTypeB>
+typename VecTypeA::elem_type IoU<useCoordinates>::Evaluate(
+    const VecTypeA& a,
+    const VecTypeB& b)
+{
+  Log::Assert(a.n_elem == b.n_elem && a.n_elem == 4, "Incorrect\
+      shape for bounding boxes. They must contain 4 elements either in the form\
+      (x0, y0, x1, y1) or (x0, y0, h, w). Refer to the documentation for more\
+      information.");
```
Comment (kartikdutt18) [inline src/mlpack/core/metrics/iou_metric_impl.hpp:30]:
```suggestion
  Log::Assert(a.n_elem == b.n_elem && a.n_elem == 4, "Incorrect \
      shape for bounding boxes. They must contain 4 elements either in the form \
      (x0, y0, x1, y1) or (x0, y0, h, w). Refer to the documentation for more \
      information.");
```

File: src/mlpack/core/metrics/iou_metric_impl.hpp [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+template<bool useCoordinates>
+template <typename VecTypeA, typename VecTypeB>
+typename VecTypeA::elem_type IoU<useCoordinates>::Evaluate(
+    const VecTypeA& a,
+    const VecTypeB& b)
+{
+  Log::Assert(a.n_elem == b.n_elem && a.n_elem == 4, "Incorrect \
+      shape for bounding boxes. They must contain 4 elements either be \
+      (x0, y0, x1, y1) or (x0, y0, h, w). Refer to the documentation \
+      for more information.");
```
Comment (KimSangYeon-DGU) [inline src/mlpack/core/metrics/iou_metric_impl.hpp:30]: Nice!, would it be a good idea to check if the coordinates consist of a rectangle?

Comment (kartikdutt18) [reply src/mlpack/core/metrics/iou_metric_impl.hpp:30]: Yes, that makes sense when coordinates are represented as `(x0, y0, x1, y1)` incase they form a line or their order isn't correct.

File: src/mlpack/core/metrics/iou_metric_impl.hpp [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+      std::min(a(0) + a(2), b(0) + b(2)) - std::max(a(0), b(0)) + 1)
+      * std::max(0.0, std::min(a(1) + a(3), b(1) + b(3)) - std::max(a(1),
+      b(1)) + 1);
+
+  return interSectionArea / (1.0 * ((a(2) + 1) * (a(3) + 1) + (b(2) + 1) *
+      (b(3) + 1) - interSectionArea));
+}
+template<bool UseCoordinates>
+template<typename Archive>
+void IoU<UseCoordinates>::serialize(Archive& ar, const unsigned int /* version */)
```
Comment (kartikdutt18) [inline src/mlpack/core/metrics/iou_metric_impl.hpp:69]:
```suggestion
void IoU<UseCoordinates>::serialize(
    Archive& ar,
    const unsigned int /* version */)
```

File: src/mlpack/tests/metric_test.cpp [hunk 5]
```diff
@@ -87,4 +97,38 @@ BOOST_AUTO_TEST_CASE(LINFMetricTest)
...
+  bbox1 << 1 << 2 << 100 << 200;
+  bbox2 << 1 << 2 << 100 << 200;
+  // IoU of same bounding boxes equals 1.0.
+  BOOST_REQUIRE_CLOSE(1.0, IoU<>::Evaluate(bbox1, bbox2), 1e-4);
+
+  // Use coordinate system to represent bounding boxes.
+  // Bounding boxes represent {x0, y0, x1, y1}.
+  bbox1 << 39 << 63 << 203 << 112;
+  bbox2 << 54 << 66 << 198 << 114;
+  // Value calculated using python interpreter.
```
Comment (KimSangYeon-DGU) [inline src/mlpack/tests/metric_test.cpp:115]: Let's change `python` to `Python` :)
```suggestion
  // Value calculated using Python interpreter.
```

File: src/mlpack/tests/metric_test.cpp [hunk 6]
```diff
@@ -87,4 +97,38 @@ BOOST_AUTO_TEST_CASE(LINFMetricTest)
...
+  BOOST_REQUIRE_CLOSE(IoU<true>::Evaluate(bbox1, bbox2), 0.7980093, 1e-4);
+
+  bbox1 << 31 << 69 << 201 << 125;
+  bbox2 << 18 << 63 << 235 << 135;
+  // Value calculated using python interpreter.
```
Comment (KimSangYeon-DGU) [inline src/mlpack/tests/metric_test.cpp:120]: Likewise.

File: src/mlpack/tests/metric_test.cpp (hunk 5 above)
Comment (kartikdutt18) [reply src/mlpack/tests/metric_test.cpp]: Awesome!, Done. Thanks a lot for the reviews.

**Comment by KimSangYeon-DGU** [comment]
Thanks, @kartikdutt18 for the awesome implementation 👍

**Comment by kartikdutt18** [comment]
Thanks @KimSangYeon-DGU for the reviews and all the help.

**Pull request merged**